
[case testTupleLowercaseSettingOff]
# flags: --python-version 3.9 --force-uppercase-builtins
x = (3,)
x = 3 # E: Incompatible types in assignment (expression has type "Literal[3]", variable has type "Tuple[builtins.int]")
[builtins fixtures/tuple.pyi]

[case testTupleLowercaseSettingOn]
# flags: --python-version 3.9 --no-force-uppercase-builtins
x = (3,)
x = 3 # E: Incompatible types in assignment (expression has type "Literal[3]", variable has type "tuple[builtins.int]")
[builtins fixtures/tuple.pyi]

[case testListLowercaseSettingOff]
# flags: --python-version 3.9 --force-uppercase-builtins
x = [3]
x = 3  # E: Incompatible types in assignment (expression has type "Literal[3]", variable has type "List[builtins.int]")

[case testListLowercaseSettingOn]
# flags: --python-version 3.9 --no-force-uppercase-builtins
x = [3]
x = 3  # E: Incompatible types in assignment (expression has type "Literal[3]", variable has type "list[builtins.int]")

[case testDictLowercaseSettingOff]
# flags: --python-version 3.9 --force-uppercase-builtins
x = {"key": "value"}
x = 3  # E: Incompatible types in assignment (expression has type "int", variable has type "Dict[str, str]")

[case testDictLowercaseSettingOn]
# flags: --python-version 3.9 --no-force-uppercase-builtins
x = {"key": "value"}
x = 3  # E: Incompatible types in assignment (expression has type "int", variable has type "dict[str, str]")

[case testSetLowercaseSettingOff]
# flags: --python-version 3.9 --force-uppercase-builtins
x = {3}
x = 3  # E: Incompatible types in assignment (expression has type "Literal[3]", variable has type "Set[builtins.int]")
[builtins fixtures/set.pyi]

[case testSetLowercaseSettingOn]
# flags: --python-version 3.9 --no-force-uppercase-builtins
x = {3}
x = 3  # E: Incompatible types in assignment (expression has type "Literal[3]", variable has type "set[builtins.int]")
[builtins fixtures/set.pyi]

[case testTypeLowercaseSettingOff]
# flags: --python-version 3.9 --no-force-uppercase-builtins
x: type[type]
y: int

y = x  # E: Incompatible types in assignment (expression has type "type[type]", variable has type "int")
