[case testTypedDictUnpackingSame]
from typing import TypedDict

class Foo(TypedDict):
    a: int
    b: int

foo1: Foo = {'a': 1, 'b': 1}
foo2: Foo = {**foo1, 'b': 2}
foo3 = Foo(**foo1, b=2)
foo4 = Foo({**foo1, 'b': 2})

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingCompatible]
from typing import TypedDict

class Foo(TypedDict):
    a: int

class Bar(TypedDict):
    a: int
    b: int

foo: Foo = {'a': 1}
bar: Bar = {**foo, 'b': 2}

[typing fixtures/typing-typeddict.pyi]



[case testTypedDictUnpackingIncompatible]
from typing import TypedDict

class Foo(TypedDict):
    a: str

class Bar(TypedDict):
    a: int
    b: int

foo: Foo = {'a': 'x'}
bar: Bar = {**foo, 'b': 2}  # E: Incompatible types (expression has type "str", TypedDict item "a" has type "int")

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingIncompatibleCorrected]
from typing import TypedDict

class Foo(TypedDict):
    a: int
    b: str

class Bar(TypedDict):
    a: int
    b: int

foo: Foo = {'a': 1, 'b': 'x'}
bar: Bar = {**foo, 'b': 2}

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingMissingKey]
from typing import TypedDict

class Foo(TypedDict):
    a: int

class Bar(TypedDict):
    a: int
    b: int

foo: Foo = {'a': 1}
bar: Bar = {**foo}  # E: Missing key "b" for TypedDict "Bar"

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingExtraKey]
from typing import TypedDict

class Foo(TypedDict):
    a: int
    c: int

class Bar(TypedDict):
    a: int
    b: int

foo: Foo = {'a': 1, 'c': 1}
bar: Bar = {**foo, 'b': 1}  # E: Extra key "c" for TypedDict "Bar"

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingNotRequiredKeyMissing]
from typing import TypedDict, NotRequired

class Foo(TypedDict):
    a: int

class Bar(TypedDict):
    a: int
    b: NotRequired[int]

foo: Foo = {'a': 1}
bar: Bar = {**foo}

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingRequiredKeyMissing]
from typing import TypedDict, NotRequired

class Foo(TypedDict):
    a: NotRequired[int]

class Bar(TypedDict):
    a: int

foo: Foo = {'a': 1}
bar: Bar = {**foo}  # E: Missing key "a" for TypedDict "Bar"

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingNotRequiredKeyIncompatible]
from typing import TypedDict, NotRequired

class Foo(TypedDict):
    a: NotRequired[str]

class Bar(TypedDict):
    a: NotRequired[int]

foo: Foo = {}
bar: Bar = {**foo}  # E: Incompatible types (expression has type "str", TypedDict item "a" has type "int")

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingNotRequiredKeyExtra]
from typing import TypedDict, NotRequired

class Foo(TypedDict):
    a: NotRequired[int]

class Bar(TypedDict):
    b: int

foo: Foo = {'a': 1}
bar: Bar = {**foo, 'b': 2}  # E: Extra key "a" for TypedDict "Bar"

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingMultiple]
from typing import TypedDict

class Foo(TypedDict):
    a: int

class Bar(TypedDict):
    b: int

class Baz(TypedDict):
    a: int
    b: int
    c: int

foo: Foo = {'a': 1}
bar: Bar = {'b': 1}
baz: Baz = {**foo, **bar, 'c': 1}

[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingNested]
from typing import TypedDict

class Foo(TypedDict):
    a: int
    b: int

class Bar(TypedDict):
    c: Foo
    d: int

foo: Foo = {'a': 1, 'b': 1}
bar: Bar = {'c': foo, 'd': 1}
bar2: Bar = {**bar, 'c': {**bar['c'], 'b': 2}, 'd': 2}

[builtins fixtures/tuple.pyi]
[builtins fixtures/dict.pyi]
[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingNestedError]
from typing import TypedDict

class Foo(TypedDict):
    a: int
    b: int

class Bar(TypedDict):
    c: Foo
    d: int

foo: Foo = {'a': 1, 'b': 1}
bar: Bar = {'c': foo, 'd': 1}
bar2: Bar = {**bar, 'c': {**bar['c'], 'b': 'wrong'}, 'd': 2}  # E: Incompatible types (expression has type "str", TypedDict item "b" has type "int")

[builtins fixtures/tuple.pyi]
[builtins fixtures/dict.pyi]
[typing fixtures/typing-typeddict.pyi]


[case testTypedDictUnpackingUntypedDict]
from typing import TypedDict

class Bar(TypedDict):
    pass

foo: dict = {}
bar: Bar = {**foo}  # E: Incompatible types (unpacked expression has type "Dict[Any, Any]", expected "_TypedDict")

[builtins fixtures/tuple.pyi]
[builtins fixtures/dict.pyi]
[typing fixtures/typing-typeddict.pyi]
