[case testDecoratedClassLine]
def d(c): ...
@d

class C: ...
class C: ...  # E: Name 'C' already defined on line 4

[case testDecoratedFunctionLine]
# flags: --disallow-untyped-defs
def d(f): ...  # type: ignore
@d

def f(): ...  # E: Function is missing a return type annotation \
              # N: Use "-> None" if function does not return a value

[case testIgnoreDecoratedFunction1]
# flags: --disallow-untyped-defs --warn-unused-ignores
def d(f): ...  # type: ignore
@d
# type: ignore
def f(): ...  # type: ignore  # E: unused 'type: ignore' comment

[case testIgnoreDecoratedFunction2]
# flags: --disallow-untyped-defs
def d(f): ...  # type: ignore
@d

def f(): ...  # type: ignore

[case testIgnoreScopeIssue1032]
def f(a: int): ...
f(
    "IGNORE"
)  # type: ignore

[case testIgnoreScopeNested1]
def f(a: int) -> int: ...
f(
    f(
        "IGNORE"
    )  # type: ignore
)

[case testIgnoreScopeNested2]
[
    "IGNORE"  # type: ignore
    &
    "IGNORE",
]
[builtins fixtures/list.pyi]

[case testIgnoreScopeNested3]
{
    "IGNORE"
    |  # type: ignore
    "IGNORE",
}
[builtins fixtures/set.pyi]

[case testIgnoreScopeNested4]
{
    None: "IGNORE"
    ^
    "IGNORE",  # type: ignore
}
[builtins fixtures/dict.pyi]

[case testIgnoreScopeNestedNonOverlapping]
def f(x: int): ...
def g(x: int): ...
(
    f("ERROR"),  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
    g("IGNORE"),  # type: ignore
    f("ERROR"),  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
)

[case testIgnoreScopeNestedOverlapping]
def f(x: int): ...
def g(x: int): ...
(
    f("ERROR"), g(  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
        "IGNORE"  # type: ignore
    ), f("ERROR"),  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
)

[case testIgnoreScopeUnused1]
# flags: --warn-unused-ignores
(  # type: ignore  # E: unused 'type: ignore' comment
    "IGNORE"  # type: ignore
    +  # type: ignore  # E: unused 'type: ignore' comment
    0  # type: ignore  # E: unused 'type: ignore' comment
)  # type: ignore  # E: unused 'type: ignore' comment

[case testIgnoreScopeUnused2]
# flags: --warn-unused-ignores
(  # type: ignore  # E: unused 'type: ignore' comment
    "IGNORE"
    -  # type: ignore
    0  # type: ignore  # E: unused 'type: ignore' comment
)  # type: ignore  # E: unused 'type: ignore' comment

[case testIgnoreScopeUnused3]
# flags: --warn-unused-ignores
(  # type: ignore  # E: unused 'type: ignore' comment
    "IGNORE"
    /
    0  # type: ignore
)  # type: ignore  # E: unused 'type: ignore' comment

-- Unskip the following PEP 570 tests once Travis supports 3.8.0a4 (or greater).

[case testPEP570ArgTypesMissing-skip]
# flags: --disallow-untyped-defs
def f(arg, /) -> None: ...  # E: Function is missing a type annotation for one or more arguments

[case testPEP570ArgTypesBadDefault-skip]
def f(arg: int = "ERROR", /) -> None: ...  # E: Incompatible default for argument "arg" (default has type "str", argument has type "int")

[case testPEP570ArgTypesDefault-skip]
def f(arg: int = 0, /) -> None:
    reveal_type(arg)  # E: Revealed type is 'builtins.int'

[case testPEP570ArgTypesRequired-skip]
def f(arg: int, /) -> None:
    reveal_type(arg)  # E: Revealed type is 'builtins.int'

[case testPEP570Required-skip]
def f(arg: int, /) -> None: ...  # N: "f" defined here
f(1)
f("ERROR")  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
f(arg=1)  # E: Unexpected keyword argument "arg" for "f"
f(arg="ERROR")  # E: Unexpected keyword argument "arg" for "f"

[case testPEP570Default-skip]
def f(arg: int = 0, /) -> None: ...  # N: "f" defined here
f()
f(1)
f("ERROR")  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
f(arg=1)  # E: Unexpected keyword argument "arg" for "f"
f(arg="ERROR")  # E: Unexpected keyword argument "arg" for "f"

[case testPEP570Calls-skip]
def f(p, /, p_or_kw, *, kw) -> None: ...  # N: "f" defined here
f(0, 0, 0)  # E: Too many positional arguments for "f"
f(0, 0, kw=0)
f(0, p_or_kw=0, kw=0)
f(p=0, p_or_kw=0, kw=0)  # E: Unexpected keyword argument "p" for "f"

[case testPEP570Signatures1-skip]
def f(p1: bytes, p2: float, /, p_or_kw: int, *, kw: str) -> None:
    reveal_type(p1)  # E: Revealed type is 'builtins.bytes'
    reveal_type(p2)  # E: Revealed type is 'builtins.float'
    reveal_type(p_or_kw)  # E: Revealed type is 'builtins.int'
    reveal_type(kw)  # E: Revealed type is 'builtins.str'

[case testPEP570Signatures2-skip]
def f(p1: bytes, p2: float = 0.0, /, p_or_kw: int = 0, *, kw: str) -> None:
    reveal_type(p1)  # E: Revealed type is 'builtins.bytes'
    reveal_type(p2)  # E: Revealed type is 'builtins.float'
    reveal_type(p_or_kw)  # E: Revealed type is 'builtins.int'
    reveal_type(kw)  # E: Revealed type is 'builtins.str'

[case testPEP570Signatures3-skip]
def f(p1: bytes, p2: float = 0.0, /, *, kw: int) -> None:
    reveal_type(p1)  # E: Revealed type is 'builtins.bytes'
    reveal_type(p2)  # E: Revealed type is 'builtins.float'
    reveal_type(kw)  # E: Revealed type is 'builtins.int'

[case testPEP570Signatures4-skip]
def f(p1: bytes, p2: int = 0, /) -> None:
    reveal_type(p1)  # E: Revealed type is 'builtins.bytes'
    reveal_type(p2)  # E: Revealed type is 'builtins.int'

[case testPEP570Signatures5-skip]
def f(p1: bytes, p2: float, /, p_or_kw: int) -> None:
    reveal_type(p1)  # E: Revealed type is 'builtins.bytes'
    reveal_type(p2)  # E: Revealed type is 'builtins.float'
    reveal_type(p_or_kw)  # E: Revealed type is 'builtins.int'

[case testPEP570Signatures6-skip]
def f(p1: bytes, p2: float, /) -> None:
    reveal_type(p1)  # E: Revealed type is 'builtins.bytes'
    reveal_type(p2)  # E: Revealed type is 'builtins.float'

[case testWalrus]
from typing import NamedTuple

if (a := 2):
    reveal_type(a)  # E: Revealed type is 'builtins.int'

while (b := "x"):
    reveal_type(b)  # E: Revealed type is 'builtins.str'

def f(x: int = (c := 4)) -> int:
    if (a := 2):
        reveal_type(a)  # E: Revealed type is 'builtins.int'

    while (b := "x"):
        reveal_type(b)  # E: Revealed type is 'builtins.str'

    x = (y := 1) + (z := 2)
    reveal_type(x)  # E: Revealed type is 'builtins.int'
    reveal_type(y)  # E: Revealed type is 'builtins.int'
    reveal_type(z)  # E: Revealed type is 'builtins.int'

    l = [y2 := 1, y2 + 2, y2 + 3]
    reveal_type(y2)  # E: Revealed type is 'builtins.int'
    reveal_type(l)  # E: Revealed type is 'builtins.list[builtins.int*]'

    filtered_data = [y3 for x in l if (y3 := a) is not None]
    reveal_type(filtered_data)  # E: Revealed type is 'builtins.list[builtins.int*]'
    reveal_type(y3)  # E: Revealed type is 'builtins.int'

    # https://www.python.org/dev/peps/pep-0572/#exceptional-cases
    (y4 := 3)
    reveal_type(y4)  # E: Revealed type is 'builtins.int'

    y5 = (y6 := 3)
    reveal_type(y5)  # E: Revealed type is 'builtins.int'
    reveal_type(y6)  # E: Revealed type is 'builtins.int'

    f(x=(y7 := 3))
    reveal_type(y7)  # E: Revealed type is 'builtins.int'

    reveal_type((lambda: (y8 := 3) and y8)())  # E: Revealed type is 'builtins.int'

    y7 = 1.0  # E: Incompatible types in assignment (expression has type "float", variable has type "int")
    if y7 := "x":  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
        pass

    # Just make sure we don't crash on this sort of thing.
    if (NT := NamedTuple("NT", [("x", int)])):  # E: "int" not callable
        z2: NT  # E: Invalid type "NT"

    if (Alias := int):
        z3: Alias  # E: Invalid type "Alias"

    return (y8 := 3) + y8

reveal_type(c)  # E: Revealed type is 'builtins.int'

[builtins fixtures/f_string.pyi]

[case testWalrusNewSemanal]
# flags: --new-semantic-analyzer
from typing import NamedTuple

if (a := 2):
    reveal_type(a)  # E: Revealed type is 'builtins.int'

while (b := "x"):
    reveal_type(b)  # E: Revealed type is 'builtins.str'

def f(x: int = (c := 4)) -> int:
    if (a := 2):
        reveal_type(a)  # E: Revealed type is 'builtins.int'

    while (b := "x"):
        reveal_type(b)  # E: Revealed type is 'builtins.str'

    x = (y := 1) + (z := 2)
    reveal_type(x)  # E: Revealed type is 'builtins.int'
    reveal_type(y)  # E: Revealed type is 'builtins.int'
    reveal_type(z)  # E: Revealed type is 'builtins.int'

    l = [y2 := 1, y2 + 2, y2 + 3]
    reveal_type(y2)  # E: Revealed type is 'builtins.int'
    reveal_type(l)  # E: Revealed type is 'builtins.list[builtins.int*]'

    filtered_data = [y3 for x in l if (y3 := a) is not None]
    reveal_type(filtered_data)  # E: Revealed type is 'builtins.list[builtins.int*]'
    reveal_type(y3)  # E: Revealed type is 'builtins.int'

    # https://www.python.org/dev/peps/pep-0572/#exceptional-cases
    (y4 := 3)
    reveal_type(y4)  # E: Revealed type is 'builtins.int'

    y5 = (y6 := 3)
    reveal_type(y5)  # E: Revealed type is 'builtins.int'
    reveal_type(y6)  # E: Revealed type is 'builtins.int'

    f(x=(y7 := 3))
    reveal_type(y7)  # E: Revealed type is 'builtins.int'

    reveal_type((lambda: (y8 := 3) and y8)())  # E: Revealed type is 'builtins.int'

    y7 = 1.0  # E: Incompatible types in assignment (expression has type "float", variable has type "int")
    if y7 := "x":  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
        pass

    # Just make sure we don't crash on this sort of thing.
    if (NT := NamedTuple("NT", [("x", int)])):  # E: "int" not callable
        z2: NT  # E: Invalid type "NT"

    if (Alias := int):
        z3: Alias  # E: Invalid type "Alias"

    return (y8 := 3) + y8

reveal_type(c)  # E: Revealed type is 'builtins.int'

[builtins fixtures/f_string.pyi]
