[case testErrorCodeAttribute]
import m
m.x  # E: Module has no attribute "x"  [attr-defined]
'x'.foobar  # E: "str" has no attribute "foobar"  [attr-defined]
[file m.py]
[builtins fixtures/module.pyi]
